examples/dcp/stateful_example.py

```diff
@@ -0,0 +1,147 @@
+
+    return loss
+
+
+def _init_model(device, world_size):
+    device_mesh = init_device_mesh(device, (world_size,))
+    model = Model().cuda()
+    model = FSDP(
+        model,
+        device_mesh=device_mesh,
+        use_orig_params=True,
+    )
+    optim = torch.optim.Adam(model.parameters(), lr=0.1)
+    _make_stateful(model, optim)
+
+    return model, optim
+
+
+def _compare_models(model1, model2, rank, rtol=1e-5, atol=1e-8):
+    model1.eval()
+    model2.eval()
+
+    with FSDP.summon_full_params(model1), FSDP.summon_full_params(model2):
+        for (name1, param1), (name2, param2) in zip(
+            model1.named_parameters(), model2.named_parameters()
+        ):
+            if name1 != name2:
+                print(f"Parameter names don't match: {name1} vs {name2}. Rank:{rank}")
+                return False
+
+            if not torch.allclose(param1, param2, rtol=rtol, atol=atol):
+                print(f"Parameters don't match for {name1}. Rank:{rank}")
+                print(
+                    f"Max difference: {(param1 - param2).abs().max().item()}. Rank:{rank}"
+                )
+                return False
+
+    print(f"All parameters match within the specified tolerance. Rank:{rank}")
+    return True
+
+
+def run(rank, world_size, region, s3_uri, device="cuda"):
+    # Set up world process group
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+
+    dist.init_process_group("cpu:gloo,cuda:nccl", rank=rank, world_size=world_size)
+    torch.cuda.set_device(rank)
+
+    print(f"Train initial model on rank:{rank}")
+    model, optim = _init_model(device, world_size)
+    _train(model, optim, train_steps=2)
+
+    print(f"Saving checkpoint on rank:{rank}")
+    # initialize S3StorageWriter with region and bucket name, before passing to dcp.save as writer
+    storage_writer = S3StorageWriter(region, s3_uri)
+    dcp.save(
+        state_dict={"model": model, "optimizer": optim},
+        storage_writer=storage_writer,
+    )
+
+    # presumably do something else
+    print(f"Train another model on rank:{rank}")
+    loaded_model, loaded_optim = _init_model(device, world_size)
+    _train(loaded_model, loaded_optim, train_steps=4)
+
+    print(f"Check that models are different on rank:{rank}")
+    assert not _compare_models(model, loaded_model, rank)
+
+    print(f"Load checkpoint on rank:{rank}")
+    #  initialize S3StorageReader with region and bucket name, before passing to dcp.load as reader
+    storage_reader = S3StorageReader(region, s3_uri)
+    dcp.load(
+        state_dict={"model": loaded_model, "optimizer": loaded_optim},
+        storage_reader=storage_reader,
+    )
+    print(f"Check that loaded model and original model are the same on rank:{rank}")
+    assert _compare_models(model, loaded_model, rank)
+
+    print(f"Train loaded model on rank:{rank}")
+    _train(loaded_model, loaded_optim, train_steps=2)
+
+    print(f"Quiting on rank:{rank}")
```

Comment:
The run method is too big, can you break it down into smaller methods for the ease of readability and understanding?